File: ruleMatch/src/main/scala/com/bigdata/rulematch/scala/old/service/impl/HBaseQueryServiceImpl.scala
package com.bigdata.rulematch.scala.old.service.impl

import com.bigdata.rulematch.scala.old.conf.EventRuleConstant
import org.apache.commons.lang3.StringUtils
import org.apache.hadoop.hbase.TableName
import org.apache.hadoop.hbase.client.{Connection, Get}
import org.apache.hadoop.hbase.util.Bytes
import org.slf4j.{Logger, LoggerFactory}

/**
 * 查询hbase的服务实现类
 */
class HBaseQueryServiceImpl {
  private val logger: Logger = LoggerFactory.getLogger(this.getClass.getName.stripSuffix("$"))

  private var hbaseConn: Connection = null

  def this(hbaseConn: Connection) = {
    this()
    this.hbaseConn = hbaseConn
  }

  /**
   * 根据给定的rowkey，查询hbase，判断查询出来的标签值与规则中的标签值是否一致
   * @param rowkey
   * @param userProfileConditions
   * @return
   */
  def userProfileConditionIsMatch(rowkey: String,
                                  userProfileConditions: Map[String, (String, String)]) = {
    var isMatch = true

    //获取到hbase的表
    val table = hbaseConn.getTable(TableName.valueOf(EventRuleConstant.HBASE_USER_PROFILE_TABLE_NAME))

    val get = new Get(Bytes.toBytes(rowkey))
    val family = "f"

    //需要查询的用户标签,只需要查询规则条件中的标签即可,不需要全部查询
    val tags = userProfileConditions.keySet
    tags.foreach(tag => {
      get.addColumn(Bytes.toBytes(family), Bytes.toBytes(tag))
    })

    // 执行get查询
    val result = table.get(get)

    //判断查询出来的标签值与用户画像规则条件中的值是否一致
    val tagsIterator = tags.iterator
    while (tagsIterator.hasNext && isMatch) {
      val tag = tagsIterator.next()
      val valueBytes: Array[Byte] = result.getValue(Bytes.toBytes(family), Bytes.toBytes(tag))

      val queryValue = Bytes.toString(valueBytes)

      val userProfileTagOptAndValue = userProfileConditions.getOrElse(tag, (EventRuleConstant.OPERATOR_EQUEAL, ""))

      val userProfileTagOpt = userProfileTagOptAndValue._1
      val userProfileTagValue = userProfileTagOptAndValue._2

      if (!compareUserTag(userProfileTagOpt, userProfileTagValue, queryValue)) {
        isMatch = false
        logger.debug(s"用户画像类规则不匹配, 标签名: ${tag}, 规则中要求的值: ${userProfileTagOptAndValue}, " +
          s"hbase中查询到的值: ${queryValue}")
      }
    }

    isMatch
  }

  def compareUserTag(opt: String, optValue: String, queryValue: String) = {
    var isMatch = false

    val optVal = StringUtils.trim(optValue)

    //是否为数字
    var optValIsNumber = false
    //对应的数字
    var optValNumber = 0L
    //查询出来的数字
    var queryValNumber = 0L

    try {
      optValNumber = optVal.toLong

      queryValNumber = StringUtils.trim(queryValue).toLong

      optValIsNumber = true
    } catch {
      case _ =>
    }

    opt match {
      case EventRuleConstant.OPERATOR_CONTAIN =>{
        isMatch = StringUtils.contains(queryValue, optVal)
      }
      case EventRuleConstant.OPERATOR_NOT_CONTAIN =>{
        isMatch = !StringUtils.contains(queryValue, optVal)
      }
      case EventRuleConstant.OPERATOR_EQUEAL => {
        isMatch = StringUtils.equals(StringUtils.trim(queryValue), optVal)
      }
      case EventRuleConstant.OPERATOR_NOT_EQUEAL => {
        isMatch = !StringUtils.equals(StringUtils.trim(queryValue), optVal)
      }
      case EventRuleConstant.OPERATOR_GREATERTHAN => {
        if(optValIsNumber){

          if(queryValue > optVal) {
            isMatch = true
          }

        }else{
          logger.error(s"规则中要求的数值 ${optVal} 或者 查询结果中的数值: ${queryValue} 不是数字类型, 不支持的比较类型: ${opt}")
        }
      }
      case EventRuleConstant.OPERATOR_GREATER_EQUEAL => {
        if(optValIsNumber){

          if(queryValue >= optVal) {
            isMatch = true
          }

        }else{
          logger.error(s"规则中要求的数值 ${optVal} 或者 查询结果中的数值: ${queryValue} 不是数字类型, 不支持的比较类型: ${opt}")
        }
      }
      case EventRuleConstant.OPERATOR_LESSTHAN => {
        if(optValIsNumber){

          if(queryValue < optVal) {
            isMatch = true
          }

        }else{
          logger.error(s"规则中要求的数值 ${optVal} 或者 查询结果中的数值: ${queryValue} 不是数字类型, 不支持的比较类型: ${opt}")
        }
      }
      case EventRuleConstant.OPERATOR_LESS_EQUEAL => {
        if(optValIsNumber){

          if(queryValue <= optVal) {
            isMatch = true
          }

        }else{
          logger.error(s"规则中要求的数值 ${optVal} 或者 查询结果中的数值: ${queryValue} 不是数字类型, 不支持的比较类型: ${opt}")
        }
      }
      case _ => {
        logger.error(s"不支持的比较类型: ${opt}")
      }
    }

    isMatch
  }
}

File: ruleMatch/src/main/java/com/bigdata/rulematch/java/old/service/HBaseQueryServiceImpl.java
package com.bigdata.rulematch.java.old.service;

import com.bigdata.rulematch.java.old.conf.EventRuleConstant;
import org.apache.commons.lang.StringUtils;
import org.apache.commons.math3.util.Pair;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.Connection;
import org.apache.hadoop.hbase.client.Get;
import org.apache.hadoop.hbase.client.Result;
import org.apache.hadoop.hbase.client.Table;
import org.apache.hadoop.hbase.util.Bytes;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Map;
import java.util.Set;

/**
 * 查询hbase的服务实现类
 *
 * @author HanDongfang
 * @version 1.0
 * @create 2021-12-21  14:17
 */
public class HBaseQueryServiceImpl {
    private Logger logger = LoggerFactory.getLogger(this.getClass().getName());

    private Connection hbaseConn = null;

    /**
     * 初始化连接
     *
     * @param hbaseConn
     */
    public HBaseQueryServiceImpl(Connection hbaseConn) {
        this.hbaseConn = hbaseConn;
    }

    /**
     * 根据给定的rowkey，查询hbase，判断查询出来的标签值与规则中的标签值是否一致
     *
     * @param rowkey
     * @param userProfileConditions
     * @return
     */
    public boolean userProfileConditionIsMatch(String rowkey, Map<String, Pair<String, String>> userProfileConditions) {
        boolean isMatch = true;

        //获取到hbase的表
        Table table = null;
        try {
            table = hbaseConn.getTable(TableName.valueOf(EventRuleConstant.HBASE_USER_PROFILE_TABLE_NAME));
            //生成get
            Get get = new Get(Bytes.toBytes(rowkey));
            //列族
            String family = "f";
            Set<String> keySet = userProfileConditions.keySet();
            //添加查询列
            for (String s : keySet) {
                get.addColumn(Bytes.toBytes(family), Bytes.toBytes(s));
            }

            // 执行get查询
            Result result = table.get(get);

            for (Map.Entry<String, Pair<String, String>> userProfileEntry : userProfileConditions.entrySet()) {
                String key = userProfileEntry.getKey();
                byte[] value = result.getValue(Bytes.toBytes(family), Bytes.toBytes(key));
                String valueStr = value.toString();

                Pair<String, String> userProfileTagOptAndValue = userProfileEntry.getValue();
                String userProfileTagOpt = userProfileTagOptAndValue.getKey();
                String userProfileValue = userProfileTagOptAndValue.getValue();

                if (!compareUserTag(userProfileTagOpt, valueStr, userProfileValue)) {
                    isMatch = false;
                    logger.debug("用户画像类规则不匹配, 标签名: {}, 规则中要求的值: {},hbase中查询到的值: {}", userProfileTagOpt, valueStr, userProfileValue);
                    break; //存在不匹配立即结束比较
                }
            }

            return isMatch;
        } catch (IOException e) {

            logger.error(String.format("获取hbase表数据出错: %s", e));
            return false;
        }
    }

    /**
     * 依据比较操作符号进行比对
     *
     * @param tagOpt
     * @param queryValueStr
     * @param ruleValueStr
     * @return
     */
    private boolean compareUserTag(String tagOpt, String queryValueStr, String ruleValueStr) {
        boolean isMatch = false;
        //是非数值
        boolean isNumber = false;
        //去掉两端空白
        queryValueStr = queryValueStr.trim();
        ruleValueStr = ruleValueStr.trim();
        int queryValue = 0;
        int ruleValue = 0;

        try {
            queryValue = Integer.valueOf(queryValueStr);
            ruleValue = Integer.valueOf(ruleValueStr);
            isNumber = true;
        } catch (NumberFormatException e) {
            logger.debug("非数值类型: ", e);
        }

        switch (tagOpt.toUpperCase()) {
            case EventRuleConstant.OPERATOR_CONTAIN:
                isMatch = StringUtils.contains(queryValueStr, ruleValueStr);
                break;

            case EventRuleConstant.OPERATOR_NOT_CONTAIN:
                isMatch = !StringUtils.contains(queryValueStr, ruleValueStr);
                break;

            case EventRuleConstant.OPERATOR_EQUEAL:
                isMatch = StringUtils.equals(queryValueStr, ruleValueStr);
                break;

            case EventRuleConstant.OPERATOR_NOT_EQUEAL:
                isMatch = !StringUtils.equals(queryValueStr, ruleValueStr);
                break;

            case EventRuleConstant.OPERATOR_GREATERTHAN:
                if (isNumber) {
                    isMatch = (queryValue > ruleValue);
                } else {
                    logger.error("非数值类型不能比较大小,queryValueStr:{},ruleValueStr{}", queryValueStr, ruleValueStr);
                }
                break;

            case EventRuleConstant.OPERATOR_GREATER_EQUEAL:
                if (isNumber) {
                    isMatch = (queryValue >= ruleValue);
                } else {
                    logger.error("非数值类型不能比较大小,queryValueStr:{},ruleValueStr{}", queryValueStr, ruleValueStr);
                }
                break;

            case EventRuleConstant.OPERATOR_LESSTHAN:
                if (isNumber) {
                    isMatch = (queryValue < ruleValue);
                } else {
                    logger.error("非数值类型不能比较大小,queryValueStr:{},ruleValueStr{}", queryValueStr, ruleValueStr);
                }
                break;

            case EventRuleConstant.OPERATOR_LESS_EQUEAL:
                if (isNumber) {
                    isMatch = (queryValue <= ruleValue);
                } else {
                    logger.error("非数值类型不能比较大小,queryValueStr:{},ruleValueStr{}", queryValueStr, ruleValueStr);
                }
                break;

            default:
                isMatch = false;
                logger.error("传入比较类型不正确: {}", tagOpt);
        }

        return isMatch;
    }
}
